### PR TITLE
feat(EDSL): narrow integer and fixed-bytes types (#2086)

### LINKED ISSUE #2086: feat(EDSL): narrow integer and fixed-bytes types (uint8–uint248, intN, bytes1–bytes31)
Child of #1724. #1 convergent gap across all three parity maps (docs/parity/): ~45 Lido hits, 22 ERC-4337 hits, ~15 Morpho hits.

## Goal
First-class sub-256-bit value types in `verity_contract`: `uintN` for N in 8..248 (notably `uint128`, `uint120`, `uint112`, `uint96`, `uint64`, `uint48`, `uint32`, `uint16`), signed `intN`, and `bytes1`–`bytes31` (notably `bytes4`, `bytes8`, `bytes16`, `bytes20`).

Required semantics:
- Narrowing casts (`uintN(x)`) with Solidity 0.8.x truncation semantics
- Params/returns/event/error payloads of narrow types (ABI encoding pads to 32 bytes)
- Arithmetic staying in the narrow domain (checked overflow at width N)
- Interaction with packed storage handled separately in #2060

## Concrete unblocks
- ERC-4337: `DepositInfo` struct, all `uint192` nonce math, `addStake(uint32)`, `unpackLow128`, `_getValidationData` (`uint48` windows)
- Lido: legacy struct compat variants, module state narrow ints
- Morpho: `uint128` market accounting fields

## Done when
- [ ] Types compile through Layer 2/3 with differential tests vs solc
- [ ] Parity map rows flip ❌→✅/🚧 and docs/parity/ is updated in the same PR

### PR DESCRIPTION
**Pull request by Th0rgal**
## Summary

- add bounded `UIntN`, two's-complement `IntN`, and fixed-width `BytesN` core carriers with Solidity-width aliases
- extend EDSL parsing, ABI signatures/layout, parameter loading, casts, events, custom errors, and external calls for `Uint24..Uint248`, `Int8..Int248`, and `Bytes1..Bytes31`
- preserve narrow expression domains and provide checked narrow arithmetic helpers with Solidity-style overflow predicates
- add a worked narrow-types contract, generated macro property test, and proof coverage across ABI/IR/Yul boundaries

## Verification

- `lake build` (2,460 jobs)
- `lake build PrintAxioms` (2,570 jobs)
- `make checks` (641 tests; zero source axioms; zero sorry)
- targeted full-proof modules exposed by `PrintAxioms` build successfully

Closes #2086

---

> [!NOTE]
> **Medium Risk**
> Large surface area in ABI encoding and macro lowering, but bounded by width validation, direct-ref event rules, and extended proof/tests. Incorrect masks or sign-extension would break on-chain compatibility rather than security boundaries.
> 
> **Overview**
> Adds Solidity-style **narrow ABI types** (`uintN`/`intN`/`bytesN`) end-to-end: new `ParamType` variants, core `UIntN`/`IntN`/`BytesN` carriers with width aliases, and compiler paths that treat them as single 32-byte words with **bit masks**, **`signextend`**, or **left-aligned `bytesN`** normalization.
> 
> The **EDSL macro** parses these types, canonicalizes literals/assignments/returns/immutables/interface returndata, routes custom-error args through typed encoding, and requires **direct parameter refs** for narrow event args. **Compile-time validation** rejects invalid widths; **`returnArray`** rules exclude narrow element arrays. **ABI JSON**, calldata loading, events, and custom errors share the new encoders; proofs and feature tests cover macro lowering and invalid-width failures.
> 
> Includes **`Contracts.Examples.NarrowTypes`**, **`narrowUInt`/`narrowInt`/`narrowBytes`**, checked **`narrow*Panic`** helpers, and external-call word encoding for narrow types. **Storage fields still reject** narrow types (deferred to #2060).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ed5c9105c8649aeb4a183278d13adbb49e11bd04. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by Th0rgal** [comment]
@codex review

File: Verity/Macro/Translate.lean [hunk 1]
```diff
@@ -1335,7 +1338,9 @@ private partial def translateDoElem
               | none =>
                   let safeBind? ← translateSafeRequireBind fields constDecls immutableDecls params locals varName rhs
                   match safeBind? with
-                  | some safeStmts => pure (safeStmts, locals.push (mkTypedLocal varName .uint256), mutableLocals)
+                  | some safeStmts =>
+                      let safeTy ← inferPureExprType fields constDecls immutableDecls externalDecls params locals rhs
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply Verity/Macro/Translate.lean]: Fixed in d508744d and retained in exact head ed5c9105c: safe require-bind typing uses inferBindSourceType, and the merged tree builds Contracts.Examples.NarrowTypes successfully.

File: Verity/Macro/Translate/Expr.lean [hunk 2]
```diff
@@ -640,12 +683,13 @@ def findContextAccessorShadowName?
...
 def isSignedWordValueType : ValueType → Bool
-  | .int256 => true
+  | .int256 | .intN _ => true
   | .newtype _ baseType => isSignedWordValueType baseType
   | _ => false
 
 def isWordLikeValueType : ValueType → Bool
-  | .uint256 | .int256 | .uint8 | .uint16 | .address | .bytes32 => true
+  | .uint256 | .int256 | .uint8 | .uint16 | .uintN _ | .intN _ | .bytesN _
+  | .address | .bytes32 => true
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply Verity/Macro/Translate/Expr.lean:692]: Fixed in exact head ed5c9105c: numeric literals compared with a BytesN peer are left-aligned before eq/!= lowering. Added deployed-Yul family regressions covering Bytes4 equality and reversed Bytes20 inequality; the NarrowTypes target builds successfully.
